### ISSUE TITLE
Illegal XML characters output in checkstyle format

### ISSUE THREAD

**Issue by Mashi** [eslint/eslint#7736]
Jenkins checkstyle complains that character reference "&#26" is an invalid XML character.  Online XML validator http://www.xmlvalidation.com/index.php?id=1&L=0 confirms.

```xml
<error line="1" column="1" severity="error" message="Parsing error: Unexpected character &apos;&#65533;&apos;&#10;&#10;&gt; 1 | &#65533;PNG&#10;    | ^&#10;  2 | &#26;&#10;  3 | &#0;&#0;&#0;&#10;  4 | IHDR&#0;&#0;&#0;F&#0;&#0;&#0;F&#8;&#6;&#0;&#0;&#0;q.&#65533;&#65533;&#0;&#0;&#0;&#4;gAMA&#0;&#0;&#65533;&#65533;&#11;&#65533;a&#5;&#0;&#0;&#0;&#9;pHYs&#0;&#0;&#14;&#65533;&#0;&#0;&#14;&#65533;&#1;&#65533;&#65533;k&#65533;&#0;&#0;&#0;&#25;tEXtSoftware&#0;Paint.NET v3.5.5I&#65533;&#65533;&#65533;&#0;&#0;" source="" /></file>
```

Environment
* ESLint Version: 3.12.0
* Node Version: 6.9.1
* npm Version: 4.0.2
* OS Version: Windows 10 x64

Using parser babel-eslint

{
  "extends": "airbnb",
  "parser": "babel-eslint"
}

On executiong:
"node ./node_modules/eslint/bin/eslint.js -f checkstyle -o lint-report.xml src/**/*

**Comment by not-an-aardvark** [comment]
Thanks for the report. Does this only occur when specific code is linted, or does it always occur?

**Comment by Mashi** [comment]
Only occurs when linting.  I would say it's a minor issue.  I know I shouldn't be linting images (binary files), I just haven't figured out how to exclude them yet.

**Comment by not-an-aardvark** [comment]
Sorry, maybe I'm misunderstanding -- you're running ESLint on an image file?

In any case, the XML formatter probably shouldn't output invalid XML, although I'm surprised anything is getting outputted other than something like "Invalid token".

**Comment by not-an-aardvark** [comment]
Oh I see, just looked at your description in the original issue above. So the problem is that ESLint isn't sanitizing parsing errors for the XML formatter.

**Comment by Mashi** [comment]
Yes, the problem is just the XML formatter output.

**Comment by platinumazure** [comment]
@Mashi Any chance you could please include the file you're linting and/or a lint report using a different formatter, so we can see more clearly what's going on?

**Comment by Mashi** [comment]
Attached the file: 
![adblogo-70](https://cloud.githubusercontent.com/assets/281359/21079352/28f1b5ac-bf90-11e6-9eb0-ad6e0d2f9c5e.png)

**Comment by not-an-aardvark** [comment]
I can't reproduce this; when I run ESLint on the image file on OSX, the resulting XML is valid.

```bash
$ node_modules/.bin/eslint not-valid-js.png -f checkstyle
<?xml version="1.0" encoding="utf-8"?><checkstyle version="4.3"><file name="/Users/tkatz/code/sandbox/not-valid-js.png"><error line="1" column="1" severity="error" message="Parsing error: Unexpected character &apos;&#65533;&apos;" source="" /></file></checkstyle>
```

Apparently `\x26` is used as an EOF character on Windows, so it's possible that this issue is windows-specific.

**Comment by BWEJesser** [comment]
I also have this problem but only when I run the lint check through Jenkins using Ant. If I run it on the command line directly everything is fine.

I'm curious how does eslint grab the list of files it will lint? There seems to be two problems, (1) eslint is trying to parse files it shouldn't and (2) the results of it parsing those files creates invalid XML when using the checkstyle format.

**Comment by not-an-aardvark** [comment]
> I also have this problem but only when I run the lint check through Jenkins using Ant. If I run it on the command line directly everything is fine.

I'm not familiar with Ant; does it just start an ESLint process and collect the output? If so, given that everything works fine on the command line, it seems like the invalid XML issue might be a bug in Ant, not in ESLint.

> I'm curious how does eslint grab the list of files it will lint?

That depends on the arguments that are passed to ESLint (you can pass it globs, files, or folders). Documentation on this can be found [here](http://eslint.org/docs/user-guide/command-line-interface).

If you give more information on what arguments you're passing to ESLint, I might be able to help figure out why it's unexpectedly trying to parse files.

**Comment by platinumazure** [comment]
@not-an-aardvark I'm not an Ant expert, but its default logger prepends every line with something and that could break XML validity if that's used as the output file.

Example similar to what I see at work (this is paraphrased though):

```
-runeslint:
[exec] Executing command: "eslint ."
[exec] D:\Data\code\path\to\file.js:
[exec] 12:16 Unexpected dangling comma.   comma-dangle
[exec] 15:51 Missing semicolon.           semi
```

If an XML formatter is used and that formatter prints newlines, then Ant will mangle the output. I think your best bet in that case is to use an output file (`-o` option on CLI) and then read that file for success/failure, rather than using anything from the process' stdout.

**Comment by BWJesseR** [comment]
Sure, so as an example here is snippet of the Ant build script, this is the piece for running the lint command:
`<target name="lint">
 <exec executable="node" output="eslint.xml">
   <arg line="node_modules/eslint/bin/eslint.js app/** -f checkstyle" />
  </exec>
</target>`

This is a React Native app so we only care about what is in the app folder and that will contain JS, images, JSON, etc. The output goes into an XML file that ends up not being valid because of the output from eslint trying to parse image files. This is the same issue @Mashi has in his original message. I get a similar result if I remove `-f checkstyle` part, in that it tries to lint images and prints out garbage characters.

From the command line I would run:
`node node_modules/eslint/bin/eslint.js app/**`

**Comment by not-an-aardvark** [comment]
It looks like you're using `app/**` as a glob, which will match every file in the `app` folder, including images, JSON, etc.

To only lint JS files, you should use something like

`node_modules/.bin/eslint app/**/*.js`

**Comment by BWJesseR** [comment]
Okay, I was getting around it by adding `--ignore-pattern *.png` but that approach looks cleaner. Any idea why I would see different results from Ant vs command line though?

In that Ant will try to parse images but the command line doesn't?

**Comment by not-an-aardvark** [comment]
I think using `node_modules/.bin/eslint app/**` will try to parse images regardless of whether you're using Ant. As @platinumazure suggested, it's possible that Ant mangles some of ESLint's output, which could cause XML to be malformed.

**Comment by BWJesseR** [comment]
Sorry, I probably didn't explain it well enough, let me try again. :)
If I run the Ant command, with the original `app/**` part, and remove the `-f checkstyle`, so I can get a count of the issues, the output from Ant will report 1213 problems. Run the same thing but from the command line there are only 100 problems.

As an example, somewhat related to this, if I run:
`node node_modules/eslint/bin/eslint.js app/shared/images/ui/checkmark.png`
I get an error with a bunch of garbage characters because it is trying to parse an image.

But if I run:
`node node_modules/eslint/bin/eslint.js app/shared/**`
It just parses the JS files it finds on that path and shows no errors from trying to parse any of the images. Shouldn't that have included the checkmark.png image and returned an error?

Either way, I recognize using `app/**/*.js` is the better approach. Just pointing out what seems like an inconsistency in what files are checked?

**Comment by not-an-aardvark** [comment]
If you're using bash, glob patterns will sometimes get expanded by bash before they ever get passed to ESLint. For example, if you run this on the bash command line:

```bash
node_modules/.bin/eslint app/shared/**
```

From ESLint's perspective, this is identical to the following:

```bash
node_modules/.bin/eslint app/shared/foo.js app/shared/bar.js app/shared/baz.png
```

...because bash will pass all the contents of `app/shared` as arguments to ESLint.

On the other hand, your Ant build script might not be expanding globs, so it's the equivalent of doing this in bash:

```bash
node_modules/.bin/eslint 'app/shared/**'
```

In other words, the glob pattern itself gets passed to ESLint when you use Ant.

Based on the [CLI docs](http://eslint.org/docs/user-guide/command-line-interface#ext), it looks like ESLint's behavior with file extensions is different depending on whether or not you pass a glob. By default, the `--ext` option is `.js`, indicating that only `.js` files are linted. However, when you pass a glob, all files in the glob are linted.

tl;dr:

* Due to the way bash processes globs, when you run `node_modules/.bin/eslint app/shared/**` in bash, no glob is "actually" passed to ESLint.
* To reproduce the Ant behavior in bash, surround the glob in quotes, e.g. `node_modules/.bin/eslint 'app/shared/**'`
* If you pass a glob to ESLint, make sure the glob specifies an extension.

**Comment by BWJesseR** [comment]
Oh, forgot to respond to @platinumazure, using the -o option and not using Ant's output results in the same issue. The invalid characters are from it trying to convert the garbage characters in the image to XML.

@not-an-aardvark thanks for that clarification. That explains the issue. 👍

**Comment by not-an-aardvark** [comment]
It looks like the conclusion was that this is not a bug in ESLint, so I'm going to close this issue.
